Turn ID into an interface

bosschaert opened the pull request:
Signed-off-by: David Bosschaert <bosschae@adobe.com>

stbischof commented:
maybe we can add a ::toString method to add a javaDoc that specifies that a Sting with the GAV-Syntax is returned.

bosschaert commented:
> maybe we can add a ::toString method to add a javaDoc that specifies that a Sting with the GAV-Syntax is returned.

Good point - I just added that.

bjhargrave commented on org.osgi.service.feature/src/org/osgi/service/feature/ID.java:
```diff
@@ -34,125 +33,53 @@
+	/**
+	 * This method returns the ID using the following syntax:
+	 * <p>
+	 * {@code groupId ':' artifactId [ ':' type [ ':' classifier ] ] ':' version }
```
Like this doc. But you should use the syntax rules from [Core 1.3.2](https://docs.osgi.org/specification/osgi.core/8.0.0/framework.introduction.html#framework.general.syntax).

`groupId ':' artifactId ( ':' type ( ':' classifier )? )? ':' version`

`[...]` means chose one from a set

stbischof commented on org.osgi.service.feature/src/org/osgi/service/feature/FeatureService.java:
```diff
@@ -35,6 +35,49 @@
+	 * @param classifier The classifier.
+	 * @return The ID.
+	 */
+	public ID getID(String groupId, String artifactId, String version,
```
Type must not be null is classifier is not null.

bosschaert replied:
Hi @stbischof there are 3 overloads for `getID()` - none of them should be used with `null` arguments but they express what is allowable through the method signature.

I also dded a constant for the Feature Type into the ID interface. I'm not sure an extra method to create IDs for features adds much value if we already have the constant.

stbischof replied:
that is enough.

stbischof replied:
last edge-case:
- What happens with empty values or if they contain a ":"?

bjhargrave replied:
For each argument, the javadoc should state whether the value must not be null or may be null. Then it is clear to the reader and the TCK implementor.

bosschaert replied:
> last edge-case:
> 
> * What happens with empty values or if they contain a ":"?

Would that not be outside the grammar definition?

    groupId ':' artifactId ( ':' type ( ':' classifier )? )? ':' version

which says: a ':' is _always_ followed by a value, or am I misunderstanding?

bjhargrave replied:
I guess it should be clear that the productions for `groupId` etc. result in non-empty tokens.

bjhargrave commented on org.osgi.service.feature/src/org/osgi/service/feature/FeatureService.java:
```diff
@@ -33,21 +33,67 @@
+	 * @param groupId The group ID (not {@code null}).
+	 * @param artifactId The artifact ID (not {@code null}).
+	 * @param version The version (not {@code null}).
+	 * @param type The type (or {@code null}).
```
If null, default is? `jar`?

bjhargrave replied:
So empty Optional and Optional of empty string mean the same thing for classifier? And empty Optional and Optional of `jar` mean the same thing for type? Otherwise how is toString to properly render the string? You don't want `::` when classifier is an Optional of empty string.

bjhargrave commented on org.osgi.service.feature/src/org/osgi/service/feature/FeatureService.java:
```diff
@@ -33,21 +33,67 @@
+	 * @param artifactId The artifact ID (not {@code null}).
+	 * @param version The version (not {@code null}).
+	 * @param type The type (or {@code null}).
+	 * @param classifier The classifier (or {@code null}).
```
If null, default is? `` (empty string)?

If non-null, doesn't type need to be non-null? Or does it default to `jar`?

bosschaert replied:
Default is that this element isn't there. I.e. there is no default.

bjhargrave replied:
As above, this is not really true.

bjhargrave commented on org.osgi.service.feature/src/org/osgi/service/feature/ID.java:
```diff
@@ -34,125 +33,58 @@
+	 * 
+	 * @return The type identifier.
+	 */
+	Optional<String> getType();
```
I am now thinking this should not be optional but instead be just String where the default value is `jar`

bosschaert replied:
No, if the type is not specified there is no type. In some contexts 'jar' is then used instead, but from an ID point of view, if it's not specified then there is no value.

bjhargrave commented on org.osgi.service.feature/src/org/osgi/service/feature/ID.java:
```diff
@@ -34,125 +33,58 @@
+	 * 
+	 * @return The classifier.
+	 */
+	Optional<String> getClassifier();
```
I am now thinking this should not be optional but instead be just String where the default value is `` (empty string)

These changes are necessary since if classifier has a value, type must also have a value. So better to just make sure they always have actual values.

bosschaert replied:
I think an alternative would be to use `null` if there is no value, but I would _not_ use an empty string.

bjhargrave commented on org.osgi.service.feature/src/org/osgi/service/feature/ID.java:
```diff
@@ -34,125 +33,58 @@
+	/**
+	 * This method returns the ID using the following syntax:
+	 * <p>
+	 * {@code groupId ':' artifactId ( ':' type ( ':' classifier )? )? ':' version }
```
To be more clear here:

- classifier is included if its value is not `` (empty string).
- type is included if its value is not `jar` or classifier's value is not `` (empty string).

bosschaert commented:
I think it's actually cleaner to simply not allow `null` to any of the ID factory methods. That way you can never have a classifier where the type is not specified. 
I would not infer any default values of type or classifier. They are either there or they are not. When the classifier is there, the type is always also there. The API expresses this.
I have pushed a small update to represent this.

bjhargrave commented:
> That way you can never have a classifier where the type is not specified.

It also means that you can never specify a type and not specify a classifier. Which means that for classifier empty Optional and Optional of empty string must be treated identically.

bjhargrave commented:
You may want to look at the api Bnd has for this: https://github.com/bndtools/bnd/blob/master/biz.aQute.repository/src/aQute/maven/api/Archive.java.

stbischof commented:
Does that mean that the ID of a Feature itsselfe always needs to have a type set?

when using it like in the spec examples
```
{
    "id": "org.acme:acmeapp:2.0.0",
      "name": "The Acme Application"
}
```
it should load a jar?

bjhargrave commented:
> Does that mean that the ID of a Feature itsselfe always needs to have a type set?

I think type must always default to `jar` if not specified. This is the maven way. GAV alone does not provide enough information about the artifact since a GAV generally has multiple artifacts (like jar, pom).

bosschaert commented:
> > That way you can never have a classifier where the type is not specified.
> 
> It also means that you can never specify a type and not specify a classifier. Which means that for classifier empty Optional and Optional of empty string must be treated identically.

No, there is a factory method that has just a type and not a classifier: https://github.com/osgi/osgi/pull/272/files#diff-0cf53bfe27354ab0be03ceead60dfc5032ecbb70bc3f44483992756309eccca4R68

bosschaert commented:
> > Does that mean that the ID of a Feature itsselfe always needs to have a type set?
> 
> I think type must always default to `jar` if not specified. This is the maven way. GAV alone does not provide enough information about the artifact since a GAV generally has multiple artifacts (like jar, pom).

If the type is not specified, the default for Maven is context-dependent. Maybe in the context of Bnd it can always be evaluated to `jar` but for example in the context of a pom the default is `pom`, e.g. see here for an example where a parent pom is referred to without explicitly declaring the type: https://github.com/apache/felix-dev/blob/master/framework/pom.xml#L22

The ID interface in features provides a general mechanism to specify an entity's ID. We should follow the Maven way of using the context to infer what the default is if it's not specified. In the context of bundles it will be `jar` but in another context it may be something else.

So, if the type is not specified when obtaining the ID, the ID should simply not contain a value for the type so it can be inferred in a context-sensitive manner when needed.

bjhargrave commented:
> he default for Maven is context-dependent.

How many contexts are their in the Feature spec? Shouldn't these be specified so that the default for type is known to all?

If you want to specify a classifier, you must specify a type. So there must be a well known default type for people to use in this case.

If I do `getID("g", "a", "1", "jar", "")` what will `toString()` return? Will it be different than `getID("g", "a", "1")`?

Part of that question is whether empty string is a valid classifier?

stbischof commented:
At the moment we use the ID for the bundles. There it will be a jar.
But we also have an id in the id attribute of the Feature it's self. Then it would be osgifeature.

Also if we will have referenced Features  in the Feature to comvine then then there may will be an additional features reference beside the bundles. Ther also would be osgifeature the default type

bosschaert commented:
> How many contexts are their in the Feature spec? Shouldn't these be specified so that the default for type is known to all?

Right now there are 2 main contexts: IDs specified for bundles and the ID specified for the feature itself. In the case of bundle the default should be `jar` and in the case of the feature the default should be `osgifeature`. I think it would be good to document this in the spec (currently it isn't).

However there could be additional different defaults for IDs used in extensions, what the default here is should be part of the extension's documentation.

> If you want to specify a classifier, you must specify a type. So there must be a well known default type for people to use in this case.

Agreed.

> If I do `getID("g", "a", "1", "jar", "")` what will `toString()` return? Will it be different than `getID("g", "a", "1")`?

    getID("g", "a", "1", "jar", "")
      -> I think this one should throw an IllegalArgument exception, an empty string classifier should not be allowed

And yes, the following should have different output IMHO:

    getID("g", "a", "1", "jar").toString() -> "g:a:jar:1"
    getID("g", "a", "1").toString() -> "g:a:1"


> Part of that question is whether empty string is a valid classifier?

I don't think an empty string is a valid argument for any of: groupID, artifactID, version, type, classifier. We should add this to the javadoc.

bosschaert commented:
I have squashed all the commits in this PR into one and updated the Javadoc with default interpretations of `type` for feature and bundle. Also I marked all factory methods for ID _not_ to allow empty values which also implies non-null.

bjhargrave commented:
> Also I marked all factory methods for ID _not_ to allow empty values which also implies non-null.

`(must not be empty)` does not also mean must not be null in that "null" is not a subset of "empty".

bosschaert commented:
> > Also I marked all factory methods for ID _not_ to allow empty values which also implies non-null.
> 
> `(must not be empty)` does not also mean must not be null in that "null" is not a subset of "empty".

I assumed that non-empty also means not null and not `""`, but I can add `not null` explicitly if that helps.

bosschaert commented:
I changed the Javadoc from `(must not be empty)` to `(not null, not empty)` and squashed everything in my latest commit.

stbischof commented on org.osgi.service.feature/src/org/osgi/service/feature/FeatureService.java:
```diff
@@ -33,21 +33,67 @@
+	 * @param mavenID The Maven ID.
+	 * @return the ID.
+	 */
+	ID getIDfromMavenID(String mavenID);
```
Is mavenId a common word? I only used Maven Coordinates in the past.

bosschaert replied:
You're right. I always called these 'Maven IDs' but looking at https://maven.apache.org/pom.html#Maven_Coordinates coordinates seems the right term. I changed the parameter name in the latest squashed commit.

bjhargrave commented on org.osgi.service.feature/src/org/osgi/service/feature/FeatureService.java:
```diff
@@ -33,21 +33,67 @@
+	 * @param coordinates The Maven Coordinates.
+	 * @return the ID.
+	 */
+	ID getIDfromMavenID(String coordinates);
```
The method name is still using "MavenID". I assumed @stbischof was commenting on the method name not just the parameter name. This method should be another overload  of `getID`.

bosschaert replied:
Sorry guys, clearly missed that one. I just fixed the method name (now `getIDfromMavenCoordinates()`), squashed and pushed again.

stbischof commented:
LGTM
